exposure-notification/src/main/java/fi/thl/covid19/exposurenotification/efgs/FederationGatewayService.java

```diff
@@ -0,0 +1,125 @@
+package fi.thl.covid19.exposurenotification.efgs;
+
+import fi.thl.covid19.exposurenotification.diagnosiskey.DiagnosisKeyDao;
+import fi.thl.covid19.exposurenotification.diagnosiskey.IntervalNumber;
+import fi.thl.covid19.exposurenotification.diagnosiskey.TemporaryExposureKey;
+import fi.thl.covid19.exposurenotification.error.EfgsOperationException;
+import fi.thl.covid19.proto.EfgsProto;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+
+import java.time.Instant;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static fi.thl.covid19.exposurenotification.efgs.FederationGatewayBatchUtil.*;
+
+@Service
+public class FederationGatewayService {
+
+    private static final int BATCH_MAX_SIZE = 5000;
+
+    private final FederationGatewayClient client;
+    private final DiagnosisKeyDao dd;
+    private final FederationOperationDao fod;
+    private final FederationBatchSigner signer;
+
+    public FederationGatewayService(
+            FederationGatewayClient client,
+            DiagnosisKeyDao diagnosisKeyDao,
+            FederationOperationDao fod,
+            FederationBatchSigner signer
+    ) {
+        this.client = client;
+        this.dd = diagnosisKeyDao;
+        this.fod = fod;
+        this.signer = signer;
+    }
+
+    public long startOutbound() {
+        long operationId = fod.startOutboundOperation();
+
+        if (operationId > 0) {
+            doOutbound(operationId);
+        }
+        return operationId;
+    }
+
+    public void startInbound(Optional<String> batchTag) {
+        String date = getDateString(Instant.now());
+        doInbound(date, batchTag);
+    }
+
+    public void startErronHandling() {
+        fod.setStalledToError();
+        List<Long> errorOperations = fod.getOutboundOperationsInError();
+        errorOperations.forEach(this::doOutbound);
+    }
+
+    // TODO: which interval should be used? Should we change interval to be more precise?
+    private void doInbound(String date, Optional<String> batchTag) {
+        client.download(date, batchTag).forEach(
+                d -> dd.addInboundKeys(transform(d), IntervalNumber.to24HourInterval(Instant.now()))
+        );
+    }
+
+    private void doOutbound(long operationId) {
+        boolean finished = false;
+        try {
+            List<List<TemporaryExposureKey>> localKeys = partitionIntoBatches(dd.fetchAvailableKeysForEfgs(operationId));
```

Comment:
Yep, that's correct and it's because efgs we''l do this batching. Also, this needs some changes to our concept, so for now it is in that way. This will be resolved later.

Reply:
Also, I think we're not gonna hit this limit very soon, even if we upload only once in a day. So, while point is not going anywhere, in practice this is just a failsafe and we should not have need to split our upload anytime soon. So, I'll get back into this later.